### PR TITLE
Add Rails 7.1 builds

### PR DESCRIPTION
**Pull request by JonRowe**

### CODE REVIEW

**Comment by pirj** [comment]
For Rails 6.1 builds:

[webdrivers 5.3.1 is installed](https://github.com/rspec/rspec-rails/actions/runs/6439621324/job/17487505003?pr=2694) for Ruby 2.6/2.7 and the build is green, but [webdrivers 5.2.0 for Ruby 3.0/3.1](https://github.com/rspec/rspec-rails/actions/runs/6439621324/job/17487504973?pr=2694) and those builds fail.

Problems are due to [this](https://github.com/titusfortner/webdrivers#update-future-of-this-project):
> If you are using an older version of webdrivers gem, and cannot upgrade, you can set the required version of chromedriver to v114 (Webdrivers::Chromedriver.required_version = '114.0.5735.90')

Where "older version" is 5.2.0, and "cannot upgrade" only affects Ruby 2.5, where webdrivers 4.x is installed.

At a glance, I can't see why 5.3.1 is not installed for all 2.6+ Ruby versions.

**Comment by taketo1113** [comment]
I think it's better to remove the webdrivers gem dependency in Ruby 3.0+.

```diff
# https://github.com/rspec/rspec-rails/blob/main/example_app_generator/generate_app.rb#L34
-  gsub_file "Gemfile", /.*chromedriver-helper.*/, "gem 'webdrivers'"
+  if RUBY_VERSION < "3.0"
+    gsub_file "Gemfile", /.*chromedriver-helper.*/, "gem 'webdrivers'"
+  end
```

Because Rails omit webdrivers gem from Gemfile template in Ruby 3.0+.
https://github.com/rails/rails/blob/main/railties/lib/rails/generators/rails/app/templates/Gemfile.tt#L64-L66
https://github.com/rails/rails/pull/48847

File: example_app_generator/generate_app.rb [hunk 1]
```diff
@@ -23,6 +23,7 @@
   gsub_file "Gemfile", /.*debug.*/, ''
   gsub_file "Gemfile", /.*puma.*/, ''
   gsub_file "Gemfile", /.*bootsnap.*/, ''
+  gsub_file "Gemfile", /.*chromedriver-helper.*/, '' if RUBY_VERSION >= "3.0" && Rails::VERSION::STRING >= '7'
```
Comment (pirj) [inline example_app_generator/generate_app.rb:26]:
```suggestion
  gsub_file "Gemfile", /.*chromedriver-helper.*/, '' if RUBY_VERSION >= "3.0" && Rails::VERSION::STRING >= '7'
  gsub_file "Gemfile", /.*webdrivers*/, '' if RUBY_VERSION >= "3.0"
```

File: example_app_generator/generate_app.rb [hunk 2]
```diff
@@ -23,6 +23,8 @@
...
+  gsub_file "Gemfile", /.*webdrivers*/, '' if RUBY_VERSION >= "3.0"
```
Comment (pirj) [inline example_app_generator/generate_app.rb:27]:
```suggestion
  gsub_file "Gemfile", /.*webdrivers.*/, '' if RUBY_VERSION >= "3.0"
```

**Comment by pirj** [comment]
I suggest skipping features on Ruby 2.5

**Comment by pirj** [comment]
For 7.1:
‘’’
Failure/Error: require 'generators/rspec'
[3204](https://github.com/rspec/rspec-rails/actions/runs/6513517460/job/17693285631?pr=2692#step:7:3205)

[3205](https://github.com/rspec/rspec-rails/actions/runs/6513517460/job/17693285631?pr=2692#step:7:3206)
NameError:
[3206](https://github.com/rspec/rspec-rails/actions/runs/6513517460/job/17693285631?pr=2692#step:7:3207)
  uninitialized constant Rails::Generators::Actions
 ‘’’

**Comment by JonRowe** [comment]
Only Rails 7 dropped webdriver, we still need it on 6.1 @pirj, also would you mind making your changes on your own branch / PR either to this branch or seperate.

**Comment by chaadow** [comment]
Hi, this kind of simple controller test is now failing for me

**Comment by chaadow** [comment]
@JonRowe  Hi this kind of test is now failing for me: 
image

with : 
```
 Failure/Error: raise TypeError, "#{path.inspect} is not a valid path: must be a String, Pathname, or Resolver"

     TypeError:
       #<RSpec::Rails::ViewRendering::EmptyTemplateResolver::ResolverDecorator:0x000000010c80fc20 @resolver=#<ActionView::FileSystemResolver:0x000000010baca2b8 @unbound_templates=#<Concurrent::Map:0x000000010baca290 entries=0 default_proc=nil>, @path_parser=#<ActionView::Resolver::PathParser:0x000000010baca218>, @path="/Users/sdsd/code/work/hoggo/app/views">> is not a valid path: must be a String, Pathname, or Resolver
```

the issue is `get :index` now fails for some reason. the solution would be to pu the entire path `get my_resource_url` But is there a way to make it work without having to edit many controller specs. Thanks!

**Comment by JonRowe** [comment]
@jcoyne 's webdriver removal actually weirdly works on 6.1 so I've picked that across with the skips for the broken specs which I'll work on seperately later to see if we can get closer to passing hre.

**Comment by JonRowe** [comment]
@chaadow This PR is about getting our CI up and running, please use the mailing list if you have problems you want to discuss or open a bug report if you think you've found an issue but remember that rspec-rails is a thin wrapper around Rails test helpers, if they don't recognise a route anymore you'll need to change.

**Comment by chaadow** [comment]
@JonRowe yes, when i updated from rspec-rails 6.0.0 to 6.0.3, the issue was solved thanks!

**Comment by chaadow** [comment]
It was fixed here https://github.com/rspec/rspec-rails/commit/bd2bb24373f2343b9d60027cf0aeb4ce791bd8df

Also I thought since the builds were failing here and i'm in the process of upgrading to rails 7.1 and had issues with rspec, that maybe the issue I have might be related to rails 7.1 👍🏼

**Comment by pirj** [comment]
> Only Rails 7 dropped webdriver

7.1 did. https://github.com/rails/rails/commit/9a532346950c74c6c745a2f22cbbb6b471a4297f 
I didn’t look if we have failures of Rails 7.0 on 3.0/3.1, it’s probably because of this.

It’s not that Rails 7.1 dropped webdriver and we are forced to use webdrivers for rails 6.1/7.0.
Existing projects running on Rails 6.1 face the same issue, and they either have to drop webdrivers gem (if they are on Ruby 3+), or use webdrivers 5.3.1 if on Ruby 2.6/2.7.

I’m unaware of a reliable solution for Rails 6.1 running on Ruby 2.5 to run browser specs.

We have to follow the same practical approach in our test suite.

I believe that my commit fixed features for Rails 6.1/7.0 running on Ruby 2.6+. I’ll be happy to send a separate PR, or @jcoyne you probably could accomodate those changes? I’m from my phone this weekend.

**Review by pirj — APPROVED** [review]
LGTM!
Please disregard my prior message.
Let’s deal with 7.1 failures and broken mailer specs separately?
Thanks!

**Comment by JonRowe** [comment]
The fixes for other builds has been merged as #2700 I'm continuing work on the 7.1 build here

**Comment by JonRowe** [comment]
Final failure was a zeitwerk exception with our generator, I've surpressed this by ignoring the file after much faffing around trying to fix the issue, it doesn't happen locally for me (but did show up in zeitwerks check) so this was a pain to debug (and is hidden in the builds by the frozen array non error).

**Pull request merged**